Repository URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID: bd3102c

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
dataset = pd.read_csv("science_data_large.csv")
print("First 15 columns:")
print(dataset.head(15))
print()
print("Summary of table information:")
print(dataset.info())
print()
print("Number of datapoints: ", len(dataset))

First 15 columns:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of table information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        10

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset.drop("Size nm^3", axis=1)
y = dataset["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame({
    "Temperature °C": 470,
    "Mols KCL": 645
}, index=[0])
sample_pred = model.predict(sample_datapoint)
print("Prediction for sample datapoint: ", sample_pred)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(x_test, y_test)
print("Model score: ", score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coeff = model.coef_
intercept = model.intercept_
print("Coefficients: ", coeff)
print("Intercept: ", intercept)

Prediction for sample datapoint:  [663785.65258473]
Model score:  0.8552472077276095
Coefficients:  [ 866.14641337 1032.69506649]
Intercept:  -409391.47958340764


Explanation for Part 3:

The R^2 score of the model is: 0.8552472077276095. This is the coefficient of determination. It provides insight into how well the model explains the variability of the target variable Size nm^3.

The R^2 score ranges from 0 to 1. A score of 1 indicates that the model perfectly fits the outcome, and a score of 0 indicates that the model does not explain any of the variability in the target variable. In this case, an R^2 value of 0.855 implies that 85.5% of the variability in the predicted variable can be explained by the features "Temperature °C" and "Mols KCl". This is considered to be a fairly good score, and it indicates that the model is robust.


<!-- Sample equation: $E = mc^2$ -->
Equation:

Let $x_1$ = Temperature °C, $x_2$ = Mols KCl, $V$ = Size in nm^3

Equation: $h(x) = V = 866.15x_1 + 1032.7x_2$ - 409390

(Rounded off to 5 significant figures according to instructions on README.md)

## Part 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_val_scores = cross_val_score(model, X, y, cv=10, scoring="r2")
print("Cross-validation score: ", cross_val_scores)
print("Mean score: ", cross_val_scores.mean())

# Report on their finding and their significance
# [My answer is in the Markdown cell below this output]

Cross-validation score:  [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Mean score:  0.8552450341984701


Analysis of Part 4:

The R^2 scores are in the range [0.783, 0.887]. This implies that the performance of the model is relatively stable across different data partitions. This means that the model generalizes well to unseen data, and is not overly fitted to the training set. The mean R^2 score is 0.855, and this implies that the model can explain about 85.5% of the variability in the target variable (Size nm^3). This means that the model captures a significant amount of information from the input features (Temperature °C and Mols KCL). Overall, the results indicate that the model is robust and reliable.

## Part 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)
poly_model = LinearRegression()
poly_model.fit(x_train_poly, y_train)

y_pred_poly = poly_model.predict(x_test_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.
score = poly_model.score(x_test_poly, y_test)
print("Model score: ", score)
coeff = poly_model.coef_
intercept = poly_model.intercept_
print("Coefficients: ", coeff)
print("Intercept: ", intercept)

Model score:  1.0
Coefficients:  [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept:  2.0477105863392353e-05


Analysis of Part 5:

The model achieved a perfect R^2 score of 1.0. This indicates that the model explains 100% of the variance in the target variable (Size nm^3). This implies that the polynomial regression model fits the training data perfectly and captures every detail without any residual error. This score is great, but it might indicate that the model is overfitted to the data. This could mean that the model does not generalize well to unseen data, meaning that it might not perform very well on the test data. The polynomial model captures interactions between the features and the quadratic relationship between them. Overall, the perfect R^2 score implies that the model has high predictive capability. 

Equation:

Let $x_1$ = Temperature °C, $x_2$ = Mols KCl, $V$ = Size in nm^3

Equation: $h(x) = V = 2.04771 \times 10^{-5} + (1.2000 \times 10^1 \times x_1) - (1.2719 \times 10^{-7} \times x_2) + (1.2649 \times 10^{-11} \times (x_1)^2) + (2.0000 \times x_1x_2) + (2.8571 \times 10^{-2} \times (x_2)^2)$